sdk/client/resourcemanager/poller_provisioning_state_helpers_test.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package resourcemanager
+
+import (
+	"encoding/json"
+	"net/http"
+	"testing"
+)
+
+// provisioningStateEndpoint is a test helper struct which makes testing the expected behaviour of the
+// provisioning state logic easier.
+//
+// Ultimately this is faking an ARM Endpoint for the Resource - as such whilst we COULD expose superfluous
+// properties (id, location, name etc) there's little point since we ignore them, so I'm intentionally not
+// doing that here, since this is "good enough" for testing purposes.
+//
+// Originally these started out as a function per test, but with so much boilerplate, this helped make the
+// tests more readable, so whilst this has some boilerplate I suspect it's worth it for the tradeoff of
+// having tests that are more understandable.
+type provisioningStateEndpoint struct {
+	expectedApiVersion  string
+	numberOfTimesCalled int
+	responses           []expectedResponse
+}
+
+func newProvisioningStateEndpoint(responses []expectedResponse) provisioningStateEndpoint {
+	return provisioningStateEndpoint{
+		expectedApiVersion:  "2020-02-01",
+		numberOfTimesCalled: 0,
+		responses:           responses,
+	}
+}
+
+func (e *provisioningStateEndpoint) endpoint(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		t.Logf("Server request %q %q", r.Method, r.RequestURI)
+
+		if r.Method != http.MethodGet && r.RequestURI != "/provisioning-state/poll" {
+			w.WriteHeader(http.StatusNotImplemented)
+			return
+		}
+		if apiv := r.URL.Query().Get("api-version"); apiv != e.expectedApiVersion {
+			w.WriteHeader(http.StatusNotImplemented)
+			t.Fatalf("expected the `api-version` to be set to `%s` but got %q", e.expectedApiVersion, apiv)
+			return
+		}
+
+		e.numberOfTimesCalled++
+		if e.numberOfTimesCalled > len(e.responses) {
+			t.Fatal("called too many times")
+		}
```

Comment:
Could we just return the last `expectedResponse` when the `responses` slice is overrun?

Reply:
We could, but since we're testing that the specific requests/responses are returning what we're expecting, I think that'd be potentially misleading in the test helper, so I'd rather not do that for now until there's a compelling use-case?

Reply:
I just thought it might more closely model the API behavior? i.e. when a final state is reached, you'd expect to see that in subsequent responses

Reply:
Depends on the API implementation - in some cases that's true (and it remains available for up to 24h, as per the ARM spec) - in some other cases it's gone once it returns `Succeeded` one-time

Reply:
Ah ok, I didn't realise that 👍